source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst

```diff
@@ -0,0 +1,287 @@
+            ]
+         ]
+      },
+      "module": "<ModuleReferenceContractVersion2>"
+   }
+
+If the module reference should be upgraded without invoking a ``migration`` function, use the below input file instead:
+
+.. code-block:: json
+
+   {
+      "migrate": {
+         "None": [
+         ]
+      },
+      "module": "<ModuleReferenceContractVersion2>"
+   }
+
+``upgrade`` function
+====================
+
+To trigger a smart contract upgrade natively on Concordium, the upgrade needs to be triggered on the ``host`` as seen in the code snippet below:
+
+.. code-block:: rust
+
+   fn contract_name<S: HasStateApi>(
+      ...
+      host: &mut impl HasHost<S>,
+      ...
+   )  ... {
+      ...
+      let result = host.upgrade(module);
+      ...
+   }
+
+.. note::
+
+   If you don't find a ``host.upgrade(module)`` snippet in a smart contract code, you can be sure that this contract
+   can not be natively upgraded. The smart contract can still be part of a ``proxy-implementation`` pattern.
+   If you want to ensure a smart contract is truly immutable you need to do your own research
+   and study its logic.
+
+The function is marked as ``low_level``. This is ``necessary`` since the high-level mutable functions store the state of the contract at the end of
+execution. This conflicts with migration since the shape of the state ``might`` be changed by the ``migration`` function. If the state is then written
+by this function it would overwrite the state stored by the ``migration`` function.
+
+Greater control is given to the smart contract developer when using ``low-level`` functions (e.g. additional state manipulation
+capabilities are exposed which are unavailable when using ``high-level`` functions). These additional state manipulation features are necessary to change the
+shape of the state during the ``migration`` but they bear the risk that the state becomes corrupted if the ``migration`` function has some coding bugs.
+Make sure you test your upgrade thoroughly (with integration tests as well as manual tests on testnet)
+and ensure your state is correctly migrated before doing
+the same smart contract upgrade on mainnet.
+
+.. note::
+
+   Using ``high-level`` function in your smart contract code (except for ``upgrade``/``migration`` functions)
+   should be your preferred option to write smart contract code safely because
+   it prevents you from accidentally corrupting the state of your smart contract. Smart contract functions are by default
+   ``high-level`` (no need to add this attribute explicitly).
+
+.. note::
+
+   You can remove the ``low-level`` attribute at the ``upgrade`` function, in other words, use a ``high-level`` function, in
+   the case that the shape of the state stays the same, in other words, no ``migration`` is executed. This means when the ``State`` struct in
+   ``contract-version1`` and ``contract-version2`` is identical, you don't need a ``migration`` function and you can remove the ``low-level`` attribute
```

Comment:
I find this paragraph a bit difficult to read. Consider revising it. 
Perhaps just removing one of the `in other words`.